File: UnbeatableTicTacToe.java
import java.util.Scanner;

public class UnbeatableTicTacToe {
    static int[][] board = new int[3][3];
    static Scanner scanner = new Scanner(System.in);
    static int HUMAN; // Human player symbol (1 for X, 2 for O)
    static int BOT;   // Bot player symbol (2 for O, 1 for X)

    public static void main(String[] args) {
        displayWelcomeMessage();
        chooseSymbol();

        int currentPlayer = HUMAN;

        while (true) {
            if (currentPlayer == HUMAN) {
                System.out.println("Your Turn! Choose a position (1-9): ");
            }

            int[] position = (currentPlayer == HUMAN) ? takeInput() : calculateBotMove();
            updateBoard(position, currentPlayer);
            displayBoard();
            if (checkWin(currentPlayer)) {
                displayBoard();
                if (currentPlayer == HUMAN) {
                    System.out.println("Congratulations! You won!");
                } else {
                    System.out.println("You can't Beat the Unbeatable BOT!");
                }
                break;
            }

            if (isGameDraw()) {
                displayBoard();
                System.out.println("It's a draw! What a tight match!");
                break;
            }

            currentPlayer = (currentPlayer == HUMAN) ? BOT : HUMAN;
        }
        displayGoodbyeMessage();
    }

    private static void displayWelcomeMessage() {
        System.out.println("*************************************");
        System.out.println("  Welcome to Tic-Tac-Toe! ");
        System.out.println("*************************************");
        System.out.println("You will play against the AI bot. Good luck!");
        System.out.println("Here’s how the board looks:");
        System.out.println("Positions: 1 | 2 | 3");
        System.out.println("           4 | 5 | 6");
        System.out.println("           7 | 8 | 9");
    }

    private static void chooseSymbol() {
        System.out.println("Choose your symbol (X or O)");
        System.out.println("Enter X for X or O for O: ");
        while (true) {
            String input = scanner.nextLine().trim().toUpperCase();
            if (input.equals("X")) {
                HUMAN = 1;
                BOT = 2;
                System.out.println("You are X. Bot is O.");
                break;
            } else if (input.equals("O")) {
                HUMAN = 2;
                BOT = 1;
                System.out.println("You are O. Bot is X.");
                break;
            } else {
                System.out.println("Invalid choice! Please choose X or O: ");
            }
        }
    }

    private static void displayGoodbyeMessage() {
        System.out.println("\n*************************************");
        System.out.println("  Thanks for playing Tic-Tac-Toe!");
        System.out.println("  See you next time!");
        System.out.println("*************************************");
    }

    private static void displayBoard() {
        System.out.println("\nCurrent Board:");
        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 3; col++) {
                String symbol = switch (board[row][col]) {
                    case 1 -> "X"; // X
                    case 2 -> "O"; // O
                    default -> " "; // Empty cell
                };
                System.out.print(" " + symbol + " ");
                if (col < 2) System.out.print("|"); // Add separator for columns
            }
            System.out.println();
            if (row < 2) System.out.println("---+---+---"); // Add separator for rows
        }
    }

    private static int[] takeInput() {
        while (true) {
            int input = scanner.nextInt() - 1;
            int row = input / 3;
            int col = input % 3;

            if (input >= 0 && input < 9 && board[row][col] == 0) {
                return new int[]{row, col};
            }
            System.out.println("Invalid move! Choose an empty position (1-9):");
        }
    }

    private static void updateBoard(int[] pos, int player) {
        board[pos[0]][pos[1]] = player;
    }

    private static boolean checkWin(int player) {
        for (int row = 0; row < 3; row++) {
            if (board[row][0] == player && board[row][1] == player && board[row][2] == player) return true;
        }

        for (int col = 0; col < 3; col++) {
            if (board[0][col] == player && board[1][col] == player && board[2][col] == player) return true;
        }

        if (board[0][0] == player && board[1][1] == player && board[2][2] == player) return true;

        return board[0][2] == player && board[1][1] == player && board[2][0] == player;
    }

    private static boolean isGameDraw() {
        for (int[] row : board) {
            for (int cell : row) {
                if (cell == 0) return false;
            }
        }
        return true;
    }

    private static int[] calculateBotMove() {
        int[] bestMove = new int[2];
        int bestScore = Integer.MIN_VALUE;

        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 3; col++) {
                if (board[row][col] == 0) {
                    board[row][col] = BOT;
                    int score = minimax(false);
                    board[row][col] = 0;

                    if (score > bestScore) {
                        bestScore = score;
                        bestMove[0] = row;
                        bestMove[1] = col;
                    }
                }
            }
        }

        System.out.println("Bot's turn...");
        return bestMove;
    }

    private static int minimax(boolean isMaximizing) {
        if (checkWin(HUMAN)) return -1;
        if (checkWin(BOT)) return 1;
        if (isGameDraw()) return 0;

        int bestScore = isMaximizing ? Integer.MIN_VALUE : Integer.MAX_VALUE;

        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 3; col++) {
                if (board[row][col] == 0) {
                    board[row][col] = isMaximizing ? BOT : HUMAN;
                    int score = minimax(!isMaximizing);
                    board[row][col] = 0;
                    bestScore = isMaximizing ? Math.max(bestScore, score) : Math.min(bestScore, score);
                }
            }
        }

        return bestScore;
    }
}

File: UnbeatableTicTacToe.cpp
#include <iostream>
#include <vector>
#include <windows.h>

using namespace std;

vector<vector<int>> board(3, vector<int>(3, 0));
int HUMAN, BOT; // Human and Bot player symbols (1 for X, 2 for O)

void displayWelcomeMessage() {
    cout << "*************************************" << endl;
    cout << "  Welcome to Tic-Tac-Toe! " << endl;
    cout << "*************************************" << endl;
    cout << "You will play against the AI bot. Good luck!" << endl;
    cout << "Here’s how the board looks:" << endl;
    cout << "Positions: 1 | 2 | 3" << endl;
    cout << "           4 | 5 | 6" << endl;
    cout << "           7 | 8 | 9" << endl;
}

void chooseSymbol() {
    string input;
    cout << "Choose your symbol (X or O)" << endl;
    cout << "Enter X for X or O for O: " << endl;
    while (true) {
        cin >> input;
        if (input == "X" || input == "x") {
            HUMAN = 1;
            BOT = 2;
            cout << "You are X. Bot is O." << endl;
            break;
        } else if (input == "O" || input == "o") {
            HUMAN = 2;
            BOT = 1;
            cout << "You are O. Bot is X." << endl;
            break;
        } else {
            cout << "Invalid choice! Please choose X or O: "<< endl;
        }
    }
}

void displayBoard() {
    cout << "\nCurrent Board:" << endl;
    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
            char symbol = ' ';
            if (board[row][col] == HUMAN) symbol = 'X';
            if (board[row][col] == BOT) symbol = 'O';
            cout << " " << symbol << " ";
            if (col < 2) cout << "|"; // Add separator for columns
        }
        cout << endl;
        if (row < 2) cout << "---+---+---" << endl; // Add separator for rows
    }
}

bool checkWin(int player) {
    for (int row = 0; row < 3; row++) {
        if (board[row][0] == player && board[row][1] == player && board[row][2] == player) return true;
    }

    for (int col = 0; col < 3; col++) {
        if (board[0][col] == player && board[1][col] == player && board[2][col] == player) return true;
    }

    if (board[0][0] == player && board[1][1] == player && board[2][2] == player) return true;

    return board[0][2] == player && board[1][1] == player && board[2][0] == player;
}

bool isGameDraw() {
    for (auto& row : board) {
        for (auto& cell : row) {
            if (cell == 0) return false;
        }
    }
    return true;
}

int minimax(bool isMaximizing) {
    if (checkWin(HUMAN)) return -1;
    if (checkWin(BOT)) return 1;
    if (isGameDraw()) return 0;

    int bestScore = isMaximizing ? -1000 : 1000;

    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
            if (board[row][col] == 0) {
                board[row][col] = isMaximizing ? BOT : HUMAN;
                int score = minimax(!isMaximizing);
                board[row][col] = 0;
                bestScore = isMaximizing ? max(bestScore, score) : min(bestScore, score);
            }
        }
    }

    return bestScore;
}

pair<int, int> calculateBotMove() {
    int bestScore = -1000;
    pair<int, int> bestMove = {-1, -1};

    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
            if (board[row][col] == 0) {
                board[row][col] = BOT;
                int score = minimax(false);
                board[row][col] = 0;

                if (score > bestScore) {
                    bestScore = score;
                    bestMove = {row, col};
                }
            }
        }
    }

    cout << "Bot's turn..." << endl;
    return bestMove;
}

pair<int, int> takeInput() {
    int input;
    while (true) {
        cout << "Your Turn! Choose a position (1-9): "<< endl;
        cin >> input;
        input -= 1; // Convert to 0-indexed

        int row = input / 3;
        int col = input % 3;

        if (input >= 0 && input < 9 && board[row][col] == 0) {
            return {row, col};
        }
        cout << "Invalid move! Choose an empty position (1-9): "<< endl;
    }
}

void updateBoard(pair<int, int> pos, int player) {
    board[pos.first][pos.second] = player;
}

void displayGoodbyeMessage() {
    cout << "\n*************************************" << endl;
    cout << "  Thanks for playing Tic-Tac-Toe!" << endl;
    cout << "  See you next time!" << endl;
    cout << "*************************************" << endl;
}

int main() {
    displayWelcomeMessage();
    chooseSymbol();

    int currentPlayer = HUMAN;

    while (true) {
        displayBoard();
        pair<int, int> position = (currentPlayer == HUMAN) ? takeInput() : calculateBotMove();
        updateBoard(position, currentPlayer);

        if (checkWin(currentPlayer)) {
            displayBoard();
            if (currentPlayer == HUMAN) {
                cout << "Congratulations! You won!" << endl;
            } else {
                cout << "Game Over !!!" << endl;
                cout << "You can't Beat the Unbeatable BOT!" << endl;
            }
            Sleep(10000);
            break;
        }

        if (isGameDraw()) {
            displayBoard();
            cout << "It's a draw! What a tight match!" << endl;
            break;
        }

        currentPlayer = (currentPlayer == HUMAN) ? BOT : HUMAN;
    }

    displayGoodbyeMessage();

    return 0;
}
